src/meta/src/barrier/checkpoint/recovery.rs

```diff
@@ -0,0 +1,406 @@
+        }
+        let database_status = self
+            .control
+            .databases
+            .get_mut(&self.database_id)
+            .expect("should exist");
+        match database_status {
+            DatabaseCheckpointControlStatus::Running(_) => {
+                let reset_request_id = 0;
+                let remaining_workers =
+                    control_stream_manager.reset_database(self.database_id, reset_request_id);
+                *database_status =
+                    DatabaseCheckpointControlStatus::Recovering(DatabaseRecoveringState {
+                        stage: DatabaseRecoveringStage::Resetting {
+                            remaining_workers,
+                            reset_workers: Default::default(),
+                            reset_request_id,
+                            backoff_future: None,
+                        },
+                        next_reset_request_id: reset_request_id + 1,
+                        retry_backoff_strategy: get_retry_backoff_strategy(),
+                    });
+            }
+            DatabaseCheckpointControlStatus::Recovering(state) => match state.stage {
+                DatabaseRecoveringStage::Resetting { .. } => {
+                    unreachable!("should not enter resetting again")
+                }
+                DatabaseRecoveringStage::Initializing { .. } => {
+                    let (backoff_future, reset_request_id) = state.next_retry();
+                    let remaining_workers =
+                        control_stream_manager.reset_database(self.database_id, reset_request_id);
+                    state.stage = DatabaseRecoveringStage::Resetting {
+                        remaining_workers,
+                        reset_workers: Default::default(),
+                        reset_request_id,
+                        backoff_future: Some(backoff_future),
+                    };
+                }
+            },
+        }
+    }
+}
+
+impl CheckpointControl {
+    pub(crate) fn on_report_failure(
+        &mut self,
+        resp: ReportDatabaseFailureResponse,
+        control_stream_manager: &mut ControlStreamManager,
+    ) -> Option<DatabaseStatusAction<'_, EnterReset>> {
+        let database_id = DatabaseId::new(resp.database_id);
+        let database_status = self.databases.get_mut(&database_id).expect("should exist");
+        match database_status {
+            DatabaseCheckpointControlStatus::Running(_) => {
+                Some(self.new_database_status_action(database_id, EnterReset))
+            }
+            DatabaseCheckpointControlStatus::Recovering(state) => match state.stage {
+                DatabaseRecoveringStage::Resetting { .. } => {
+                    // ignore reported failure during resetting or backoff.
+                    None
+                }
+                DatabaseRecoveringStage::Initializing { .. } => {
+                    warn!(database_id = database_id.database_id, "");
```

Comment:
After meta sends out a `ResetDatabaseRequest`, it will ignore all responses until receiving the `ResetDatabaseResponse`, and then enter the `Initializing` stage. In other word, we can assume that after we pass the `Resetting` stage, there won't be any inflight message of the failed database.

Reply:
> we can assume that after we pass the Resetting stage, there won't be any inflight message of the failed database.

Why? In my example, all `ResetDatabaseResponse` are received before the stale `ReportDatabaseFailureResponse` arrives at meta.

Reply:
Discussed offline, my example is not possible because Meta-CN uses a bidi-streaming gRPC and ordering is guaranteed, which means if meta receives `ResetDatabaseResponse` from CN, all preceding messages including `ReportDatabaseFailureResponse` must already be received by meta as well.